content/browser/attribution_reporting/attribution_storage_sql.cc

```diff
@@ -2999,6 +3388,272 @@ bool AttributionStorageSql::StoreAttributionReport(AttributionReport& report) {
+      "WHERE F.origin=? "
+      "AND F.epoch IN (";
+
+  auto& items = attribution_epochs;
+  if (!items.empty()) {
+    std::copy(items.begin(), items.end() - 1, std::ostream_iterator<uint64_t>(oss, ", "));
+    oss << items.back();
+  }
+  oss << ")";
+
+  sql::Statement statement(db_.GetUniqueStatement(oss.str().c_str()));
+  statement.BindString(0, net::SchemefulSite(querying_origin).Serialize());
+
+  base::flat_map<uint64_t, double> remaining_budgets;
+  while (statement.Step()) {
+      remaining_budgets[statement.ColumnInt64(0)] = statement.ColumnDouble(1);
+  }
+
+  // For epochs not already stored in the DB initialize their
+  // remaining budget and mark them to be inserted now
+  std::vector<uint64_t> epochs_to_insert;
+  std::vector<uint64_t> epochs_to_update;
+  for (auto& epoch : attribution_epochs) {
+    auto it = remaining_budgets.find(epoch);
+    if (it == remaining_budgets.end()) {
+      epochs_to_insert.push_back(epoch);
+      remaining_budgets[epoch] = kInitialBudget;
+    } else {
+      epochs_to_update.push_back(epoch);
+    }
+  }
+  
+  // -------------------------------------------------------
+  LOG(INFO) << "Epochs to insert:";
+  for (auto epoch : epochs_to_insert) {
+    LOG(INFO) << epoch;
+  }
+  LOG(INFO) << "Epochs to update:";
+  for (auto epoch : epochs_to_update) {
+    LOG(INFO) << epoch;
+  }
+  
+  LOG(INFO) << "Remaining Budgets:";
+  for (const auto& pair : remaining_budgets) {
+    LOG(INFO) << "- Epoch " << pair.first << " Budget " << pair.second;      
+  }
+// -------------------------------------------------------
+
+  // Check if all remaining budgets are enough for required_budget
+  for (const auto& pair : remaining_budgets) {
+    if (pair.second < required_budget) {
+      return AggregatableResult::kInsufficientBudget;
+    }
+  }
+
+  // Insert epochs not already stored
+  static constexpr char kInsertEpochOrigin[] =
+    "INSERT INTO per_origin_filters"
+    "(epoch,origin,initial_budget,consumed_budget)"
+    "VALUES(?,?,?,?)";
+  
+  for (auto& epoch : epochs_to_insert) {
+    sql::Statement stmt(db_.GetCachedStatement(SQL_FROM_HERE, kInsertEpochOrigin));
+    stmt.BindInt64(0, epoch);
+    stmt.BindString(1, net::SchemefulSite(querying_origin).Serialize());
+    stmt.BindDouble(2, kInitialBudget);
+    stmt.BindDouble(3, required_budget);
+    if (!stmt.Run()) {
+      return AggregatableResult::kInternalError;
+    }  
+  }
+  
+  // Update epochs already stored
+  static constexpr char kAdjustBudgetConsumed[] =
+    "UPDATE per_origin_filters "
+    "SET consumed_budget=consumed_budget+?"
+    "WHERE epoch=? and origin=?";
+
+  for (auto& epoch : epochs_to_update) {
+    sql::Statement stmt(db_.GetCachedStatement(SQL_FROM_HERE, kAdjustBudgetConsumed));
+    stmt.BindDouble(0, required_budget);
+    stmt.BindInt64(1, epoch);
+    stmt.BindString(2, net::SchemefulSite(querying_origin).Serialize());
+    if (!stmt.Run()) {
+      return AggregatableResult::kInternalError;
+    }
+  }
```

Comment:
How do the "?" symbols work here, is it a sort of f-string that you can execute with this `sql::Statement` thing? That looks neater than the sketchy string concatenation at least

Reply:
This is a typical SQL template I guess. I don't know exactly how the binding works internally, it's somehow replacing the ? symbols with the values I bind them with. It is neater yes! I hate the string concatenation I did too